File: vuln-sandbox/go/cwe-918/ssrf-safe.go
// SAFE: go-http-get-sprintf — URL validated against allowlist before outbound request
// Rule: go-http-get-sprintf | CWE-918 | Expected: TrueNegative

package main

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
)

var allowedHosts = map[string]bool{
	"api.example.com":  true,
	"data.example.com": true,
}

func proxyHandler(w http.ResponseWriter, r *http.Request) {
	targetURL := r.URL.Query().Get("url")

	// SAFE: parse and validate the URL against an allowlist of trusted hosts
	parsed, err := url.Parse(targetURL)
	if err != nil || !allowedHosts[parsed.Hostname()] {
		http.Error(w, "URL not allowed", http.StatusBadRequest)
		return
	}

	resp, err := http.Get(parsed.String())
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Fprintf(w, "%s", body)
}

func main() {
	http.HandleFunc("/proxy", proxyHandler)
	http.ListenAndServe(":8080", nil)
}

File: vuln-sandbox/rust/cwe-918/ssrf_safe.rs
// SAFE: rust-command-web-input — URL validated against allowlist before outbound request
// Rule: rust-command-web-input | CWE-918 | Expected: TrueNegative

use std::collections::HashSet;

async fn fetch_allowed_url(target_url: &str) -> Result<String, String> {
    let allowed_hosts: HashSet<&str> = ["api.example.com", "data.example.com"].iter().cloned().collect();

    // SAFE: parse and validate the URL against an allowlist of trusted hosts
    let parsed = url::Url::parse(target_url).map_err(|e| e.to_string())?;
    let host = parsed.host_str().ok_or("No host in URL")?;

    if !allowed_hosts.contains(host) {
        return Err(format!("Host '{}' is not allowed", host));
    }

    let response = reqwest::get(parsed).await.map_err(|e| e.to_string())?;
    response.text().await.map_err(|e| e.to_string())
}

#[tokio::main]
async fn main() {
    let url = std::env::args().nth(1).unwrap_or_else(|| "https://api.example.com/data".to_string());
    match fetch_allowed_url(&url).await {
        Ok(body) => println!("{}", body),
        Err(e) => eprintln!("Error: {}", e),
    }
}
